### ISSUE TITLE
select to copy does not work over the bondaries of WSLg

### ISSUE THREAD

**Issue by Brettdah** [microsoft/wslg#642]
# Environment

```none
Windows build number: 10.0.22000.0
Your Distribution version: 20.04
Your WSLg version: 1.0.26
```

# Steps to reproduce

On WSL distro :
- Install the terminal : Terminator
- Launch it ;)
- On Terminator : Use keyboard shortcut to split the windows (Ctrl + shift + E, split verticaly or Ctrl + shift + O, split horizontaly)
- Select any text on one of of the splitted terminal
- On the same splitted terminal use the middle mouse click to paste the selected text
- On the other one see that it work too.
- On a windows app paste it (for me it does not work, but if I right click and copy it work...)

# Actual behavior

I have to, right click and on this menu click copy too actualy copy the text over windows applications.

#  Expected behavior

- If I select something in my teminal, it should be "shared" in the windows clipboard too so I can paste it on windows appllications...
NB : this work with VcXsrv and I want WSLg to replace my need for that tool.

Thank you :)

**Comment by hideyukn88** [comment]
@Brettdah, thanks for reporting the issue. Yes, currently PRIMARY selection in X11 is not shared with outside of X world, that said, it is not just cross WSLg boundary to Windows side, but it is also not shared with Wayland native application. Thanks!

**Comment by Brettdah** [comment]
Hi I found that to copy from Windows  to linux I can do it with CTRL + Shift + V 
A good explaination about that : [here](https://superuser.com/questions/421463/why-does-ctrl-v-not-paste-in-bash-linux-shell/421468)

**Comment by Michael-Sinz** [comment]
The problem I have is that the X11 clipboard and the Windows clipboard in WSL2 on Windows 11 do not interoperate.

In Windows 10 I had to run the VcXsrv X11 server on Windows in order to get X11 applications to display on Windows.  This does have the clipboard connected such that I can highlight in one xterm window and paste in a Windows window (and the reverse too)

With Windows 11 and the built-in X11 support, this does not work.

**Comment by hideyukn88** [comment]
@Michael-Sinz, currently WSLg only support text (utf-8) for X11 and Windows clipboard exchange, or do you observe the problem with text? For wayland-native applications, WSLg does support limited HTML and bitmap formats in addition to text. Thanks!

**Comment by Michael-Sinz** [comment]
The problem is with simple ascii text - but it is on the primary selection which seems to be the problem.

**Comment by ianzhang366** [comment]
I'm having the same issue as well.

Copy from x11 app to windows works fine. 
Copy from windows to x11 won't work.

With VcXsrv(PRIMARY selection enabled) + wsl, copy between windows and x11 app works fine.

I wonder if there's a workaround to config the wslg to make sure x11 app and window clipboard know each other.

**Comment by michael-jedlicka** [comment]
I had the same issue after upgrading, used to work before but then stopped after Windows upgrade to 11. I found that in terminator the option to "Copy on Selection" was not selected which resolved my issue

**Comment by Michael-Sinz** [comment]
I have been successfully using the following in my system.  Within WSL, I run a build of autocutsel (https://github.com/sigmike/autocutsel) with the following:

```
others=$(ps a | grep auto[c]utsel | awk '{print $1}')
[[ -n ${others} ]] && kill -9 ${others} 2>/dev/null
$(dirname $0)/autocutsel -s CLIPBOARD &
$(dirname $0)/autocutsel -s PRIMARY   -buttonup &
```

The reason I don't use the `-fork` option is that I want it connected to the console/window/process for lifetime management and for any error message output that may happen.  (This is a window I normally run minimized but if there are problems, I can check on it - it has some other bits in it too)

I also have it kill prior versions just in case (easiest way to restart it all is to run the script again)

This syncs the clipboard with the primary selection but will only update the primary selection on button up (mainly to prevent clipboard change storms)

I have been using this for over 6 months now and it has served me well.

**Comment by smprather** [comment]
> I have been successfully using the following in my system. Within WSL, I run a build of autocutsel (https://github.com/sigmike/autocutsel) with the following:
> I have been using this for over 6 months now and it has served me well.

Thanks @Michael-Sinz . This is the only thing that has worked for me so far! I have confirmed bi-directional operation with no glitches / down-sides so far.

**Comment by FredSRichardson** [comment]
I set up `autocutsel` as a user `systemd` process and it seems to launch correctly.  The problem I'm having is that copy/paste still doesn't work reliably to/from an `xterm`.  It does seem to work to/from my X11 `emacs` though.

**Comment by Michael-Sinz** [comment]
> I have been successfully using the following in my system. Within WSL, I run a build of autocutsel (https://github.com/sigmike/autocutsel) with the following:
> 
> ```
> others=$(ps a | grep auto[c]utsel | awk '{print $1}')
> [[ -n ${others} ]] && kill -9 ${others} 2>/dev/null
> $(dirname $0)/autocutsel -s CLIPBOARD &
> $(dirname $0)/autocutsel -s PRIMARY   -buttonup &
> ```
> 
> The reason I don't use the `-fork` option is that I want it connected to the console/window/process for lifetime management and for any error message output that may happen. (This is a window I normally run minimized but if there are problems, I can check on it - it has some other bits in it too)
> 
> I also have it kill prior versions just in case (easiest way to restart it all is to run the script again)
> 
> This syncs the clipboard with the primary selection but will only update the primary selection on button up (mainly to prevent clipboard change storms)
> 
> I have been using this for over 6 months now and it has served me well.

One problem is that sometimes it will not copy a second time in a given direction until you copy (does not need to paste) from the other side.

What I see is that I sometimes don't get a new item from the Linux rxvt/xterm window into Windows but after I select and copy into the clipboard something different on the Windows side I can again copy something new from the Linux side.

I have not had a chance to track down the problem - but I think it is an issue of who thinks who is up-to-date but it is easily worked around.

I still would want to have the X11 support built into Windows working better.

I have actually started switching back to https://github.com/marchaesen/vcxsrv (which has continued to work rather well, even on Windows 11 and WSL2, albeit you have to configure it in special ways)

**Comment by dermoth** [comment]
I have held back to VcXsrv for a long time as well. I just tried on a new VM install, some of the issues are now fixed (ex. non-US keyboard support, dead keys...) but this would be probably a big one if I was intending to use it a lot.

With VcXsrv both PRIMARY and CLIPBOARD works natively within X (I can have different selection in them), but also both update the Windows keyboard (i.e. the last one to be used). This is IMHO the desired behavior. The Windows keyboard IIRC also updates both (IIRC - I almost exclusively use middle mouse to paste within X so I could be wrong).

Another small annoyance is the borders of WSLg, bug #530 - VcXsrv looks much nicer.

**Comment by Michael-Sinz** [comment]
I agree - and have been using VcXsrv for 8+ years now.   The current version from GitHub (marchaesen/vcxsrv: Windows X-server<https://github.com/marchaesen/vcxsrv>) is rather stable and works well with my workflow.

(I being able to just throw new windows from scripts on the Linux box and have either terminals or data displays show up rather than having to have full desktop setup)

Then again, I am an old Unix person and X11 is automatic in my mouse usage.  (middle-click paste, and all)


__

𝕄𝕚𝕔𝕙𝕒𝕖𝕝 𝕊𝕚𝕟𝕫 – Architect – 緑 – Microsoft

________________________________
